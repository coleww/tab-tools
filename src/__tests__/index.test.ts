import { getNote, getUniqueNotes, validateTabData } from '../';

describe('getNote', () => {
  it('gets open string', () => {
    expect(getNote('a', '0')).toBe('a');
  });
  it('gets note', () => {
    expect(getNote('a', '11')).toBe('g#');
  });
  it('handles octave', () => {
    expect(getNote('a', '12')).toBe('a');
  });

  it('handles frets above 12', () => {
    expect(getNote('a', '13')).toBe('a#');
  });

  it('wraps around notes array', () => {
    expect(getNote('g#', '1')).toBe('a');
  });

  it('translates flats to sharps', () => {
    expect(getNote('ab', '0')).toBe('g#');
  });

  it('handles casing', () => {
    expect(getNote('C#', '0')).toBe('c#');
  });

  it('ignores non-note tab chars', () => {
    expect(getNote('g#', 'h')).toBe('');
  });

  it('ignores bad root notes', () => {
    expect(getNote('z', '0')).toBe('');
  });
});

describe('validateTabData', () => {
  const mockTabData = {
    tuning: ['g', 'd', 'a', 'e'],
    data: [
      ['2', '', '1', ''],
      ['3', '', '2', ''],
      ['3', '', '2', ''],
      ['1', '', '0', ''],
    ],
  };

  it('returns true for valid data', () => {
    expect(validateTabData(mockTabData)).toBe(true);
  });

  it('returns true for single string input', () => {
    expect(
      validateTabData({
        tuning: ['e'],
        data: [['1', '3', '5']],
      })
    ).toBe(true);
  });

  it('detects tuning mismatch', () => {
    expect(
      validateTabData({
        ...mockTabData,
        tuning: ['b', 'f', 'd'],
      })
    ).toBe(false);
  });

  it('detects string length mismatch', () => {
    expect(
      validateTabData({
        ...mockTabData,
        data: [...mockTabData.data.slice(1), ['', '', '', '', '']],
      })
    ).toBe(false);
  });
});

describe('getUniqueNotes', () => {
  const mockTabData = {
    tuning: ['g', 'd', 'a', 'e'],
    data: [
      ['2', '', '1', ''],
      ['3', '', '2', ''],
      ['3', '', '2', ''],
      ['1', '', '0', ''],
    ],
  };

  it('returns unique notes for tab data', () => {
    expect(getUniqueNotes(mockTabData)).toStrictEqual(['a', 'b', 'c', 'e', 'f', 'g#']);
  });
});