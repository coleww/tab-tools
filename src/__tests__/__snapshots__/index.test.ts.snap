// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keyLookup matches snapshot 1`] = `
{
  "a#bc#d#ef#g#": [
    "b maj",
    "g# min",
  ],
  "a#bc#d#egg#": [
    "g# harmMin",
  ],
  "a#bc#d#ff#g#": [
    "d# min",
    "f# maj",
  ],
  "a#bc#d#fgg#": [
    "g# melMin",
  ],
  "a#bc#def#g": [
    "b harmMin",
  ],
  "a#bc#def#g#": [
    "b melMin",
  ],
  "a#bc#defg": [
    "a# oct1",
  ],
  "a#bc#defgg#": [
    "b oct2",
    "d oct2",
    "f oct2",
    "g# oct2",
  ],
  "a#bc#degg#": [
    "g oct1",
  ],
  "a#bc#efgg#": [
    "e oct1",
  ],
  "a#bdd#ff#g#": [
    "d# harmMin",
  ],
  "a#c#d#f#g#": [
    "d# pentMin",
    "f# pentMaj",
  ],
  "a#c#d#fg#": [
    "a# pentMin",
    "c# pentMaj",
  ],
  "a#c#defgg#": [
    "c# oct1",
  ],
  "a#cc#d#ef#g#": [
    "c# melMin",
  ],
  "a#cc#d#ff#g#": [
    "a# min",
    "c# maj",
  ],
  "a#cc#d#fgg#": [
    "f min",
    "g# maj",
  ],
  "a#cc#efgg#": [
    "f harmMin",
  ],
  "a#cd#fg": [
    "c pentMin",
    "d# pentMaj",
  ],
  "a#cd#fg#": [
    "f pentMin",
    "g# pentMaj",
  ],
  "a#cdd#ff#g#": [
    "d# melMin",
  ],
  "a#cdd#fgg#": [
    "c min",
    "d# maj",
  ],
  "a#cdef#g#": [
    "a# wholeTone",
    "c wholeTone",
    "d wholeTone",
    "e wholeTone",
    "f# wholeTone",
    "g# wholeTone",
  ],
  "a#cdefgg#": [
    "f melMin",
  ],
  "a#cdfg": [
    "a# pentMaj",
    "g pentMin",
  ],
  "aa#bcc#dd#eff#gg#": [
    "chromatic",
  ],
  "aa#c#defg": [
    "d harmMin",
  ],
  "aa#cc#d#ef#": [
    "a oct1",
  ],
  "aa#cc#d#ef#g": [
    "a# oct2",
    "c# oct2",
    "e oct2",
    "g oct2",
  ],
  "aa#cc#d#f#g": [
    "f# oct1",
  ],
  "aa#cc#d#ff#": [
    "a# harmMin",
  ],
  "aa#cc#d#fg": [
    "a# melMin",
  ],
  "aa#cd#ef#g": [
    "d# oct1",
  ],
  "aa#cdd#f#g": [
    "g harmMin",
  ],
  "aa#cdd#fg": [
    "a# maj",
    "g min",
  ],
  "aa#cdef#g": [
    "g melMin",
  ],
  "aa#cdefg": [
    "d min",
    "f maj",
  ],
  "abc#d#ef#g": [
    "e melMin",
  ],
  "abc#d#ef#g#": [
    "c# min",
    "e maj",
  ],
  "abc#d#ff#g#": [
    "f# melMin",
  ],
  "abc#d#fg": [
    "a wholeTone",
    "b wholeTone",
    "c# wholeTone",
    "d# wholeTone",
    "f wholeTone",
    "g wholeTone",
  ],
  "abc#def#g": [
    "b min",
    "d maj",
  ],
  "abc#def#g#": [
    "a maj",
    "f# min",
  ],
  "abc#defg": [
    "d melMin",
  ],
  "abc#dff#g#": [
    "f# harmMin",
  ],
  "abc#ef#": [
    "a pentMaj",
    "f# pentMin",
  ],
  "abcd#ef#g": [
    "e harmMin",
  ],
  "abcdd#ff#g#": [
    "a oct2",
    "c oct2",
    "d# oct2",
    "f# oct2",
  ],
  "abcdd#fg": [
    "c melMin",
  ],
  "abcdd#fg#": [
    "g# oct1",
  ],
  "abcdef#g": [
    "e min",
    "g maj",
  ],
  "abcdef#g#": [
    "a melMin",
  ],
  "abcdefg": [
    "a min",
    "c maj",
  ],
  "abcdefg#": [
    "a harmMin",
  ],
  "abcdff#g#": [
    "f oct1",
  ],
  "abdd#ff#g#": [
    "d oct1",
  ],
  "abdef#": [
    "b pentMin",
    "d pentMaj",
  ],
  "abdeg": [
    "e pentMin",
    "g pentMaj",
  ],
  "acc#d#ef#g": [
    "c oct1",
  ],
  "acc#d#ef#g#": [
    "c# harmMin",
  ],
  "acdeg": [
    "a pentMin",
    "c pentMaj",
  ],
  "acdfg": [
    "d pentMin",
    "f pentMaj",
  ],
  "bc#d#f#g#": [
    "b pentMaj",
    "g# pentMin",
  ],
  "bc#ef#g#": [
    "c# pentMin",
    "e pentMaj",
  ],
  "bcdd#ff#g#": [
    "b oct1",
  ],
  "bcdd#fgg#": [
    "c harmMin",
  ],
}
`;

exports[`keyMap matches snapshot 1`] = `
{
  "a harmMin": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g#",
  ],
  "a maj": [
    "a",
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "a melMin": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "a min": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
  ],
  "a oct1": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
  ],
  "a oct2": [
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "a pentMaj": [
    "a",
    "b",
    "c#",
    "e",
    "f#",
  ],
  "a pentMin": [
    "a",
    "c",
    "d",
    "e",
    "g",
  ],
  "a wholeTone": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "a# harmMin": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "f#",
  ],
  "a# maj": [
    "a",
    "a#",
    "c",
    "d",
    "d#",
    "f",
    "g",
  ],
  "a# melMin": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "a# min": [
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "a# oct1": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
  ],
  "a# oct2": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "a# pentMaj": [
    "a#",
    "c",
    "d",
    "f",
    "g",
  ],
  "a# pentMin": [
    "a#",
    "c#",
    "d#",
    "f",
    "g#",
  ],
  "a# wholeTone": [
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "b harmMin": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g",
  ],
  "b maj": [
    "a#",
    "b",
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
  ],
  "b melMin": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "b min": [
    "a",
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g",
  ],
  "b oct1": [
    "b",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "b oct2": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
    "g#",
  ],
  "b pentMaj": [
    "b",
    "c#",
    "d#",
    "f#",
    "g#",
  ],
  "b pentMin": [
    "a",
    "b",
    "d",
    "e",
    "f#",
  ],
  "b wholeTone": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "c harmMin": [
    "b",
    "c",
    "d",
    "d#",
    "f",
    "g",
    "g#",
  ],
  "c maj": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f",
    "g",
  ],
  "c melMin": [
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
    "g",
  ],
  "c min": [
    "a#",
    "c",
    "d",
    "d#",
    "f",
    "g",
    "g#",
  ],
  "c oct1": [
    "a",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "c oct2": [
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "c pentMaj": [
    "a",
    "c",
    "d",
    "e",
    "g",
  ],
  "c pentMin": [
    "a#",
    "c",
    "d#",
    "f",
    "g",
  ],
  "c wholeTone": [
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "c# harmMin": [
    "a",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
  ],
  "c# maj": [
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "c# melMin": [
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
  ],
  "c# min": [
    "a",
    "b",
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
  ],
  "c# oct1": [
    "a#",
    "c#",
    "d",
    "e",
    "f",
    "g",
    "g#",
  ],
  "c# oct2": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "c# pentMaj": [
    "a#",
    "c#",
    "d#",
    "f",
    "g#",
  ],
  "c# pentMin": [
    "b",
    "c#",
    "e",
    "f#",
    "g#",
  ],
  "c# wholeTone": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "chromatic": [
    "a",
    "a#",
    "b",
    "c",
    "c#",
    "d",
    "d#",
    "e",
    "f",
    "f#",
    "g",
    "g#",
  ],
  "d harmMin": [
    "a",
    "a#",
    "c#",
    "d",
    "e",
    "f",
    "g",
  ],
  "d maj": [
    "a",
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g",
  ],
  "d melMin": [
    "a",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
  ],
  "d min": [
    "a",
    "a#",
    "c",
    "d",
    "e",
    "f",
    "g",
  ],
  "d oct1": [
    "a",
    "b",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "d oct2": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
    "g#",
  ],
  "d pentMaj": [
    "a",
    "b",
    "d",
    "e",
    "f#",
  ],
  "d pentMin": [
    "a",
    "c",
    "d",
    "f",
    "g",
  ],
  "d wholeTone": [
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "d# harmMin": [
    "a#",
    "b",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "d# maj": [
    "a#",
    "c",
    "d",
    "d#",
    "f",
    "g",
    "g#",
  ],
  "d# melMin": [
    "a#",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "d# min": [
    "a#",
    "b",
    "c#",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "d# oct1": [
    "a",
    "a#",
    "c",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "d# oct2": [
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "d# pentMaj": [
    "a#",
    "c",
    "d#",
    "f",
    "g",
  ],
  "d# pentMin": [
    "a#",
    "c#",
    "d#",
    "f#",
    "g#",
  ],
  "d# wholeTone": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "e harmMin": [
    "a",
    "b",
    "c",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "e maj": [
    "a",
    "b",
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
  ],
  "e melMin": [
    "a",
    "b",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "e min": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f#",
    "g",
  ],
  "e oct1": [
    "a#",
    "b",
    "c#",
    "e",
    "f",
    "g",
    "g#",
  ],
  "e oct2": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "e pentMaj": [
    "b",
    "c#",
    "e",
    "f#",
    "g#",
  ],
  "e pentMin": [
    "a",
    "b",
    "d",
    "e",
    "g",
  ],
  "e wholeTone": [
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "f harmMin": [
    "a#",
    "c",
    "c#",
    "e",
    "f",
    "g",
    "g#",
  ],
  "f maj": [
    "a",
    "a#",
    "c",
    "d",
    "e",
    "f",
    "g",
  ],
  "f melMin": [
    "a#",
    "c",
    "d",
    "e",
    "f",
    "g",
    "g#",
  ],
  "f min": [
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "g",
    "g#",
  ],
  "f oct1": [
    "a",
    "b",
    "c",
    "d",
    "f",
    "f#",
    "g#",
  ],
  "f oct2": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
    "g#",
  ],
  "f pentMaj": [
    "a",
    "c",
    "d",
    "f",
    "g",
  ],
  "f pentMin": [
    "a#",
    "c",
    "d#",
    "f",
    "g#",
  ],
  "f wholeTone": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "f# harmMin": [
    "a",
    "b",
    "c#",
    "d",
    "f",
    "f#",
    "g#",
  ],
  "f# maj": [
    "a#",
    "b",
    "c#",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "f# melMin": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "f# min": [
    "a",
    "b",
    "c#",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "f# oct1": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "f#",
    "g",
  ],
  "f# oct2": [
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
    "f#",
    "g#",
  ],
  "f# pentMaj": [
    "a#",
    "c#",
    "d#",
    "f#",
    "g#",
  ],
  "f# pentMin": [
    "a",
    "b",
    "c#",
    "e",
    "f#",
  ],
  "f# wholeTone": [
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
  "g harmMin": [
    "a",
    "a#",
    "c",
    "d",
    "d#",
    "f#",
    "g",
  ],
  "g maj": [
    "a",
    "b",
    "c",
    "d",
    "e",
    "f#",
    "g",
  ],
  "g melMin": [
    "a",
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g",
  ],
  "g min": [
    "a",
    "a#",
    "c",
    "d",
    "d#",
    "f",
    "g",
  ],
  "g oct1": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "g",
    "g#",
  ],
  "g oct2": [
    "a",
    "a#",
    "c",
    "c#",
    "d#",
    "e",
    "f#",
    "g",
  ],
  "g pentMaj": [
    "a",
    "b",
    "d",
    "e",
    "g",
  ],
  "g pentMin": [
    "a#",
    "c",
    "d",
    "f",
    "g",
  ],
  "g wholeTone": [
    "a",
    "b",
    "c#",
    "d#",
    "f",
    "g",
  ],
  "g# harmMin": [
    "a#",
    "b",
    "c#",
    "d#",
    "e",
    "g",
    "g#",
  ],
  "g# maj": [
    "a#",
    "c",
    "c#",
    "d#",
    "f",
    "g",
    "g#",
  ],
  "g# melMin": [
    "a#",
    "b",
    "c#",
    "d#",
    "f",
    "g",
    "g#",
  ],
  "g# min": [
    "a#",
    "b",
    "c#",
    "d#",
    "e",
    "f#",
    "g#",
  ],
  "g# oct1": [
    "a",
    "b",
    "c",
    "d",
    "d#",
    "f",
    "g#",
  ],
  "g# oct2": [
    "a#",
    "b",
    "c#",
    "d",
    "e",
    "f",
    "g",
    "g#",
  ],
  "g# pentMaj": [
    "a#",
    "c",
    "d#",
    "f",
    "g#",
  ],
  "g# pentMin": [
    "b",
    "c#",
    "d#",
    "f#",
    "g#",
  ],
  "g# wholeTone": [
    "a#",
    "c",
    "d",
    "e",
    "f#",
    "g#",
  ],
}
`;
